### ISSUE TITLE
Incorrect optional chaining tree-shaking

### ISSUE THREAD

**Issue by sapphi-red** [rollup/rollup#4806]
### Rollup Version

3.10.0

### Operating System (or Browser)

macos / browser

### Link To Reproduction

https://rollupjs.org/repl/?version=3.10.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHZhbHVlJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMGdldCh2KSUyMCU3QiU1Q24lMjAlMjBsZXQlMjBnJTIwJTNEJTIwdmFsdWUlM0YuZmluZCgoYiklMjAlM0QlM0UlMjB2JTIwJTNEJTNEJTNEJTIwYiklM0IlNUNuJTIwJTIwaWYlMjAoIWcpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwcmV0dXJuJTIwZy5jaGFyQXQoMCklM0IlNUNuJTdEJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHNldChhKSUyMCU3QiU1Q24lMjAlMjB2YWx1ZSUyMCUzRCUyMGElM0IlNUNuJTdEJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

### Expected Behaviour

The output `get` function returns the found value.

```js
export function get(v) {
  let g = value?.find((b) => v === b);
  if (!g) {
    return;
  }
  return g.charAt(0);
}
```

### Actual Behaviour

The output `get` function always returns `undefined`.
```js
export function get(v) {
  value?.find((b) => v === b);
  {
    return;
  }
}
```

This tree-shaking can't be done because `value` can be set by `set` function and can be a non-undefined value.

Original issue: https://github.com/vitejs/vite/issues/11715

**Comment by lukastaegert** [comment]
Ah I see. If you add a call `set()` to the module itself, it works correctly, so apparently the external export does not receive the same treatment here. Will need to look into that. Probably a bug in the deoptimization logic. Another way to "fix" it is to switch the order of the exports, which again points to a deoptimization bug. https://rollupjs.org/repl/?version=3.10.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHZhbHVlJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHNldChhKSUyMCU3QiU1Q24lMjAlMjB2YWx1ZSUyMCUzRCUyMGElM0IlNUNuJTdEJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMGdldCh2KSUyMCU3QiU1Q24lMjAlMjBsZXQlMjBnJTIwJTNEJTIwdmFsdWUlM0YuZmluZCgoYiklMjAlM0QlM0UlMjB2JTIwJTNEJTNEJTNEJTIwYiklM0IlNUNuJTIwJTIwaWYlMjAoIWcpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwcmV0dXJuJTIwZy5jaGFyQXQoMCklM0IlNUNuJTdEJTVDbiU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

**Comment by satoren** [comment]
It appears that an early return before using optional chaining will "fix" it too.
https://rollupjs.org/repl/?version=3.10.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHZhbHVlJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMGdldCh2KSUyMCU3QiU1Q24lMjAlMjBpZiUyMCghdmFsdWUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHJldHVybiUzQiU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwbGV0JTIwZyUyMCUzRCUyMHZhbHVlJTNGLmZpbmQoKGIpJTIwJTNEJTNFJTIwdiUyMCUzRCUzRCUzRCUyMGIpJTNCJTVDbiUyMCUyMGlmJTIwKCFnKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjByZXR1cm4lM0IlNUNuJTIwJTIwJTdEJTVDbiUyMCUyMHJldHVybiUyMGcuY2hhckF0KDApJTNCJTVDbiU3RCU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjBzZXQoYSklMjAlN0IlNUNuJTIwJTIwdmFsdWUlMjAlM0QlMjBhJTNCJTVDbiU3RCU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

**Comment by lukastaegert** [comment]
Fix at #4812.
